### ISSUE TITLE
[red-knot] Either support metaclasses that are not subclasses of `type`, or emit a diagnostic when encountering them

### ISSUE THREAD

**Issue by AlexWaygood** [astral-sh/ruff#14208]
Believe it or not, Python accepts arbitrary callables as metaclasses in `class` statements -- and those callables don't have to return instances of `type`. As long as the callable accepts the same arguments as `type.__new__`, it's an acceptable metaclass; the binding created by the class statement is set to whatever calling the metaclass returns:

```pycon
>>> def returns_int(*args: object) -> int:
...     return 42
...     
>>> class A(metaclass=returns_int): ...
... 
>>> A
42
```

Currently we don't understand this in red-knot: we think that the `class A` statement constitutes a class definition, when it does not. Mypy ["solves" this problem](https://mypy-play.net/?mypy=latest&python=3.12&gist=f27cb48434f3e879d42cb54615444755) by emitting a diagnostic saying that a dynamic metaclass like this isn't supported. Pyright doesn't emit a diagnostic, and [just gets it wrong](https://pyright-play.net/?strict=true&code=CYUwZgBATiAuCuUB2BnA%2BgSybAFAKgEMoBzFALggHsAjAKxAGNYBKCAWgD4ItYyAoCIOhxESCABYATHz4MANgRQoIAQRwBbOAXmKUAXhgJk6HswoA6SzJgA3EATlpYATwAOIHCuZ8gA).

I think mypy's "solution" here is acceptable (metaclasses that are not subclasses of `type` are really very rare). But it also might not be too much work for us to just add an accurate understanding of the runtime semantics here, and infer that the `class A` statement binds the `A` symbol to an instance of `int` rather than a newly created class.

**Comment by AlexWaygood** [comment]
> Pyright's behavior seems to be a deliberate design choice:

Thanks, that's interesting! I still dislike pyright's behaviour here, though, even if it's deliberate. There's lots of potentially invalid metaclasses that pyright simply doesn't detect, and that I think it would be quite easily to statically detect.
